### PR TITLE
Pyzhou/tdi 43097 snowflake eliminate shutdown hook

### PR DESCRIPTION
**Pull request by pyzhou-talend**
https://jira.talendforge.org/browse/TDI-43097

### CODE REVIEW

File: jdbc/src/main/java/org/talend/components/jdbc/output/Output.java [hunk 1]
```diff
@@ -120,6 +121,9 @@ public void preDestroy() {
         if (datasource != null) {
             datasource.close();
         }
+        if ("Snowflake".equals(configuration.getDataset().getConnection().getDbType())) {
```
Comment (wwang-talend) [inline jdbc/src/main/java/org/talend/components/jdbc/output/Output.java]: 100% execution? no need "try finally" code?

Comment (pyzhou-talend) [reply jdbc/src/main/java/org/talend/components/jdbc/output/Output.java]: All exceptions will be just catch and log inside the method.

Comment (wwang-talend) [reply jdbc/src/main/java/org/talend/components/jdbc/output/Output.java]: if "datasource.close()" failed, the code will not be executed, if that is considered, it is ok

**Comment by rmannibucau** [comment]
The lack of thread safety will lead to concurrent exceptions making the delzetion missed.
I also wonder why the component *instance* creating the temp files is not rezponsible to delzte them. With big data disteibution it can happen it is not on the same node so deletion will not be done as well.
Another nasty case is in the remote engine where the static tracker can be shared by jobs so you will delete too early the file for a job.
Finally, this kind of logic should move to a service with a key per component in a tracking map IMHO, static should be almost forbidden in component to ensure a fast initialization and to avoid leaking state (except for constants).
A trivial workaround can be a service cleaning that in predestroy and ensure no trigger/ui action use temp files.

Hope it makes sense

**Comment by pyzhou-talend** [comment]
Thanks, Romain, for the review, Didn't consider these concurrency issues before. I will fix them accordingly.

File: jdbc/src/main/java/org/talend/components/jdbc/output/statement/operations/snowflake/SnowflakeCopy.java [hunk 2]
```diff
@@ -93,7 +88,7 @@ private static Path createWorkDir() {
         try {
             final Path tmp = createTempDirectory("talend-jdbc-snowflake-");
             log.debug("Temp folder {} created.", tmp);
-            tmp.toFile().deleteOnExit();
+            tmpFiles.add(tmp);
```
[the remark this answers is not in the document]
Comment (pyzhou-talend) [reply jdbc/src/main/java/org/talend/components/jdbc/output/statement/operations/snowflake/SnowflakeCopy.java]: Sorry, I didn't quite get this. Where should I get the ID or should I create it? and what service should include this?

[the remark this answers is not in the document]
Comment (rmannibucau) [reply jdbc/src/main/java/org/talend/components/jdbc/output/statement/operations/snowflake/SnowflakeCopy.java]: You generate it
And was thinking to @Service
Main difference with storing that in the component is you can have a predestroy naturally
That said the best in general is to clean asap temp files (as soon as read).

So the points and overall goals are:

1. Put any logic in a @Service classes, nothing static
2. The lifecycle of temp files must be explicit (when is it created, written, read, deleted)
3. Temp files are only allowed in the scope of a single component *instance* (avoid distribution issue)
4. Temp files must be bound to a single component instance (avoid concurrent job issues), this is where an id can help depending on the codestyle/solution you pick

Hope it makes sense

[the remark this answers is not in the document]
Comment (pyzhou-talend) [reply jdbc/src/main/java/org/talend/components/jdbc/output/statement/operations/snowflake/SnowflakeCopy.java]: The explanation is very explicit. I think I get it now. Thanks a lot.

**Pull request merged**